### PR TITLE
prefixing ability for resource requests

### PR DESCRIPTION
**Pull request by waxenegger**
see: https://trello.com/c/faOotWHI/246-support-prefix

the prefix is now taken into account, so that if I set 'omero.web.prefix' to '/me/fancy/prefix' urls use it.

prefix

image

Test:
Check on cowfish using open-with iviewer or by using the url: https://cowfish.openmicroscopy.org/webtrial/omero_iviewer/fill_in_existing_image_id

One way to test is to open your browser's developer tools (e.g. chrome: F12) and check the network tab to see if there are any requests for static or dynamic resources failing, specifically with 404 errors.

### CODE REVIEW

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [iviewer-merge#8](https://ci.openmicroscopy.org/job/iviewer-merge/8/). See the [console output](https://ci.openmicroscopy.org/job/iviewer-merge/8/consoleText) for more details.
Possible conflicts:
  - PR #6 waxenegger 'title supposed to show iViewer: image name'
    - src/app/header.js
    - src/utils/constants.js

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [iviewer-merge#10](https://ci.openmicroscopy.org/job/iviewer-merge/10/). See the [console output](https://ci.openmicroscopy.org/job/iviewer-merge/10/consoleText) for more details.
Possible conflicts:
  - PR #4 waxenegger 'cache buster'
    - plugin/omero_iviewer/views.py

File: ol3-viewer/src/ome/ol3/source/Image.js [hunk 1]
```diff
@@ -154,14 +154,13 @@ ome.ol3.source.Image = function(options) {
         console.error("The given server information is invalid!");
 
 	/**
- 	 * the omero server address/ip
+ 	 * the uri to use (incl. a porential prefix)
```
Comment (waxenegger) [inline ol3-viewer/src/ome/ol3/source/Image.js]: yep. typo.

**Comment by will-moore** [comment]
You're generating ```/webclient``` and ```/webgateway/``` urls in slightly different ways, using PREFIX for webclient and ```reverse()``` for webgateway, so you end up with this generated in index.html template:
```
        window.INITIAL_REQUEST_PARAMS = {};
            window.INITIAL_REQUEST_PARAMS['IMAGE_ID'] = "3732";
            window.INITIAL_REQUEST_PARAMS['WEBGATEWAY'] = "/test/will/webgateway/";
            window.INITIAL_REQUEST_PARAMS['URI_PREFIX'] = "/test/will";
            window.INITIAL_REQUEST_PARAMS['WEBCLIENT'] = "/test/will/webclient";
```
Seems best to use just one approach that generates urls with/without trailing slash.

**Comment by waxenegger** [comment]
The generated params are correct. Unfortunately I can only work with what I can extract. I can do a reverse lookup for the webgateway root but not webclient. The information that I use for the prefix is correct as well (SCRIPT_NAME) which is what I use for webclient.

This is just the entry point where I don't bother with the trailing slash. I have the appropriate read and getPrefixedURI methods which make sure that I get the uris in a way the application expects it to concatenate further.

**Comment by will-moore** [comment]
Why can't you reverse lookup webclient root?  ```reverse('webindex')``` should work.

**Comment by will-moore** [comment]
General comment (not a blocker): using ES6 string formatting for generating urls would be much nicer than all the ```+``` concatenations.

**Comment by waxenegger** [comment]
I shall be damned. I looked at urls.py and found webgateway but did not see the webclient definition. In that case I'm going to use reverse lookup of course.

File: src/app/header.js [hunk 2]
```diff
@@ -136,6 +137,7 @@ export class Header extends EventSubscriber {
              let url =
                 Misc.assembleImageLink(
                     this.context.server,
+                    this.context.getPrefixedURI(WEBCLIENT),
```
Comment (will-moore) [inline src/app/header.js]: Does this now have a trailing slash using ```reverse('webindex`)```, which is not handled?
assembleImageLink does:
```
let url = server + prefixed_uri + "/img_detail/" + image_id + '/?';
```

Comment (waxenegger) [inline src/app/header.js]: internally getPrefixedURI uses Misc.prepareURI which 'prepares' it such that I have slash in the beginning and no trailing slash

Comment (will-moore) [inline src/app/header.js]: OK, yep - I see.

**Comment by will-moore** [comment]
All seems to be working fine on cowfish. Didn't see any errors / 404s etc.
Looks good to merge.

**Comment by jburel** [comment]
Merging so we can "unlock" the other PRs.

**Pull request merged**